doc/command_ref.rst

```diff
@@ -1599,36 +1596,53 @@ Upgrade-To Command
 Specifying Packages
 ===================
 
-Many commands take a ``<package-spec>`` parameter that selects a package for the
-operation. DNF looks for interpretations of the parameter from the most commonly
-used meanings to the least, that is it tries to see if the given spec fits one
-of the following patterns (in decreasing order of priority):
+Many commands take a ``<package-name-spec>`` parameter that selects a package
+for the operation. DNF looks for interpretations of the parameter from the most
+specific NEVRA form to the least specific one, trying to see if the given spec
+fits one of the following forms (in decreasing order of priority):
 
-* ``name.arch``
-* ``name``
 * ``name-[epoch:]version-release.arch``
 * ``name-[epoch:]version-release``
 * ``name-[epoch:]version``
+* ``name.arch``
+* ``name``
```

Comment:
The original order and even yours is incorrect. Please see: HY_FORMS_MOST_SPEC in libdnf

Reply:
Hmm, you're right. It's confusing, because the enums that define the forms are in a different order (the one I used), and that order actually seems logical, at least at first glance. The real order doesn't seem to make much sense, so the reasons for it should be documented, maybe even explained in the man page.

Do you know the reasons for why the order is the way it is?

Reply:
The order is base on feedback and reported issue from users. The rest is based on suggestions to improve performance. We know that full NEVRA is excellent for parsing therefore it gets a priority. name must be before name-version because it caused problems. There was a package like `gimp-2` and `gimp` in version. People more often specify name that name with version.

Reply:
Hmm.. doesn't seem very scientific. I'll just fix the order in the man page...

To finish the discussion, the order seems a bit arbitrary. I think if you put N before NEV, you should also put NEV before NEVR, for the same reason: `gimp-2-2` should be NV, not NVR (that is assuming you have a version = 2 and release = 2, which your example also assumes for the version and which I think is practically very improbable...).

So, in theory, based on the above, we could perhaps actually set the order from the least specific to the most specific:
N
NA
NEV
NEVR
NEVRA

Because what really matters (because it's ambiguous) is the relation of N, NEV, and NEVR, right? The rest is really not (unless you get really creative with the architecture...). Anyway, just a theory :slightly_smiling_face:

Reply:
I would not recommend to change current default order. Probably because we increased restrictiveness of parsing, therefore there could be not negative effect.